File: 6KYU/Roman Numerals Decoder/solution.clj
(ns translate-roman-numerals.core)

(def conversions {\I 1 \V 5 \X 10 \L 50 \C 100 \D 500 \M 1000})

(defn translate-roman-numerals [roman]
  (loop [numerals roman
         last nil
         sum 0]
    (let [current (first numerals)]
      (if (empty? numerals)
        sum
        (cond (and (= current \V) (= last \I)) (recur (rest numerals) current (+ sum 3))
              (and (= current \X) (= last \I)) (recur (rest numerals) current (+ sum 8))
              (and (= current \L) (= last \X)) (recur (rest numerals) current (+ sum 30))
              (and (= current \C) (= last \X)) (recur (rest numerals) current (+ sum 80))
              (and (= current \D) (= last \C)) (recur (rest numerals) current (+ sum 300))
              (and (= current \M) (= last \C)) (recur (rest numerals) current (+ sum 800))
              :else                            (recur (rest numerals) current (+ sum (conversions current))))))))

File: 6KYU/Roman Numerals Decoder/solution.go
package kata

func Decode(roman string) int {
	conversions := map[rune]int{
		'I': 1,
		'V': 5,
		'X': 10,
		'L': 50,
		'C': 100,
		'D': 500,
		'M': 1000}

	l := len(roman)
	last := rune(roman[0])
	res := conversions[last]

	for i := 1; i < l; i++ {
		c := rune(roman[i])
		if c == 'V' && last == 'I' {
			res += 3
		} else if c == 'X' && last == 'I' {
			res += 8
		} else if c == 'L' && last == 'X' {
			res += 30
		} else if c == 'C' && last == 'X' {
			res += 80
		} else if c == 'D' && last == 'C' {
			res += 300
		} else if c == 'M' && last == 'C' {
			res += 800
		} else {
			res += conversions[c]
		}
		last = c
	}
	return res
}
